// @nous-research/ui/src/ui/components/watchlist.tsx
'use client'

import { cn } from '../../utils'

import { Scramble } from './ascii'
import { LinkIcon } from './icons'
import { Typography } from './typography'

const ETH_RE = /^0x[a-fA-F0-9]{40}$/
const truncate = (a: string) => `${a.slice(0, 6)}${'·'.repeat(8)}${a.slice(-4)}`

export function Watchlist({
  className,
  counter = false,
  items,
  scramble = false,
  ...props
}: WatchlistProps) {
  return (
    <div className={cn('flex flex-col gap-3', className)} {...props}>
      {items.map(({ label, right, url }, i) => {
        const isStr = typeof label === 'string'
        const eth = isStr && ETH_RE.test(label)
        const text = eth ? truncate(label) : (label as string)

        return (
          <a
            className={cn(
              'grid items-center gap-2.5 px-2.5 py-1.5',
              'text-display leading-[1.4]',
              'hover:bg-midground/10! hover:ring-2 hover:ring-current/20',
              'transition-all duration-500 hover:duration-0',
              'opacity-(--midground-alpha)'
            )}
            href={url}
            key={i}
            rel="noopener noreferrer"
            style={{
              background: `color-mix(in oklch, var(--color-midground) ${10 * Math.max(0, 1 - i / 9)}%, transparent)`,
              gridTemplateColumns: [
                counter && 'auto auto',
                '1fr',
                right && 'auto',
                url && 'auto auto'
              ]
                .filter(Boolean)
                .join(' ')
            }}
            target="_blank"
          >
            {counter && (
              <>
                <Typography
                  className="text-lg tracking-[0.35em] opacity-40"
                  compressed
                >
                  {String(i + 1).padStart(2, '0')}
                </Typography>

                <span className="text-[0.8125rem] font-bold tracking-[0.4em] opacity-20">
                  :
                </span>
              </>
            )}

            {isStr ? (
              <Typography
                className="min-w-0 overflow-hidden text-lg font-bold tracking-[0.35em]"
                {...(eth ? { mono: true } : { compressed: true })}
              >
                {scramble ? <Scramble delay={i * 80} text={text} /> : text}
              </Typography>
            ) : (
              label
            )}

            {right && (
              <Typography
                className="text-right text-sm tracking-widest opacity-40"
                mono
              >
                {right}
              </Typography>
            )}

            {url && (
              <>
                <span className="text-[0.8125rem] tracking-[0.4em] opacity-20">
                  :
                </span>
                <LinkIcon className="text-midground size-3.5" />
              </>
            )}
          </a>
        )
      })}
    </div>
  )
}

interface WatchlistProps extends React.ComponentProps<'div'> {
  counter?: boolean
  items: { label?: React.ReactNode; right?: React.ReactNode; url?: string }[]
  scramble?: boolean
}
