// @nous-research/ui/dist/ui/header.js
"use client";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { AnimatePresence, motion } from "motion/react";
import { createElement, useCallback, useRef, useState } from "react";
import { useCssVarDims } from "../hooks/use-css-var-dims.js";
import { useGpuTier } from "../hooks/use-gpu-tier.js";
import { cn } from "../utils/index.js";
import { Blink } from "./components/blink.js";
import { Cell, Grid } from "./components/grid/index.js";
import { HoverBg } from "./components/hover-bg.js";
import { HamburgerIcon } from "./components/icons/hamburger.js";
import { Scramble } from "./components/scramble.js";
import { Socials } from "./components/socials.js";
import { ThemeToggle } from "./components/theme-toggle.js";
import { H2 } from "./components/typography/h2.js";
import { Small } from "./components/typography/small.js";
const DEFAULT_BRAND = /* @__PURE__ */ jsxs("hgroup", { className: "flex flex-col gap-2", children: [
  /* @__PURE__ */ jsx(Small, { children: "Nous" }),
  /* @__PURE__ */ jsx(H2, { children: "Research" })
] });
const DEFAULT_LINKS = [
  { href: "/projects", label: "Projects" },
  { href: "/participants", label: "Participants" },
  { href: "/provenance", label: "Provenance" },
  { href: "/contribute", label: "Contribute" }
];
export function Header({
  brand = DEFAULT_BRAND,
  brandHref = "/",
  className,
  desktopGridStyle,
  links = DEFAULT_LINKS,
  LinkComponent = "a",
  scramble: scrambleProp = true,
  socials,
  socialsLabel = "Socials",
  style,
  themeLabel = "Theme",
  themeToggle = false
}) {
  const ref = useRef(null);
  useCssVarDims("header", ref);
  const gpuTier = useGpuTier();
  const scramble = scrambleProp && gpuTier > 0;
  const [open, setOpen] = useState(false);
  const close = useCallback(() => setOpen(false), []);
  const hasSocials = (socials?.length ?? 0) > 0;
  const hasMobileChrome = themeToggle || hasSocials;
  return /* @__PURE__ */ jsxs("header", { className, ref, style, children: [
    /* @__PURE__ */ jsxs(
      Grid,
      {
        className: "hidden border-t border-b lg:grid",
        style: desktopGridStyle,
        children: [
          /* @__PURE__ */ jsx(
            BrandCell,
            {
              brand,
              href: brandHref,
              LinkComponent
            }
          ),
          links.map((link) => /* @__PURE__ */ jsx(
            NavCell,
            {
              link,
              LinkComponent,
              scramble
            },
            link.href
          )),
          hasSocials && /* @__PURE__ */ jsxs(Cell, { className: "flex items-start justify-between", children: [
            /* @__PURE__ */ jsx(Small, { className: "opacity-50", children: socialsLabel }),
            /* @__PURE__ */ jsx(Socials, { items: socials })
          ] }),
          themeToggle && /* @__PURE__ */ jsxs(Cell, { className: "flex items-start justify-between", children: [
            /* @__PURE__ */ jsx(Small, { className: "opacity-50", children: themeLabel }),
            /* @__PURE__ */ jsx(ThemeToggle, {})
          ] })
        ]
      }
    ),
    /* @__PURE__ */ jsxs(
      "div",
      {
        className: cn(
          "flex items-center justify-between border border-current/20 p-4",
          "lg:hidden"
        ),
        children: [
          /* @__PURE__ */ jsx(
            BrandLink,
            {
              brand,
              href: brandHref,
              LinkComponent
            }
          ),
          /* @__PURE__ */ jsxs("div", { className: "flex items-center gap-3", children: [
            themeToggle && /* @__PURE__ */ jsx(ThemeToggle, {}),
            /* @__PURE__ */ jsx(
              "button",
              {
                "aria-label": open ? "Close menu" : "Open menu",
                className: "relative z-50 cursor-pointer bg-transparent p-2",
                onClick: () => setOpen((v) => !v),
                type: "button",
                children: /* @__PURE__ */ jsx(HamburgerIcon, { open })
              }
            )
          ] })
        ]
      }
    ),
    /* @__PURE__ */ jsx(AnimatePresence, { children: open && /* @__PURE__ */ jsx(
      motion.div,
      {
        animate: { opacity: 1 },
        className: cn(
          "bg-background/95 fixed inset-0 z-50 flex flex-col backdrop-blur-sm",
          "p-8 lg:hidden"
        ),
        exit: { opacity: 0 },
        initial: { opacity: 0 },
        transition: { duration: 0.2 },
        children: /* @__PURE__ */ jsxs("div", { className: "flex flex-col border border-current/20", children: [
          /* @__PURE__ */ jsxs("div", { className: "flex items-center justify-between border-b border-current/20 p-4", children: [
            /* @__PURE__ */ jsx(
              BrandLink,
              {
                brand,
                href: brandHref,
                LinkComponent,
                onClick: close
              }
            ),
            /* @__PURE__ */ jsx(
              "button",
              {
                "aria-label": "Close menu",
                className: "cursor-pointer bg-transparent p-2",
                onClick: close,
                type: "button",
                children: /* @__PURE__ */ jsx(HamburgerIcon, { open: true })
              }
            )
          ] }),
          links.map((link) => /* @__PURE__ */ jsx(
            MobileNavLink,
            {
              link,
              LinkComponent,
              onNavigate: close,
              scramble
            },
            link.href
          )),
          hasMobileChrome && /* @__PURE__ */ jsxs("div", { className: "flex items-center gap-3 border-b border-current/20 p-4", children: [
            hasSocials && /* @__PURE__ */ jsxs(Fragment, { children: [
              /* @__PURE__ */ jsx(Small, { className: "opacity-50", children: socialsLabel }),
              /* @__PURE__ */ jsx(Socials, { items: socials, onNavigate: close })
            ] }),
            themeToggle && hasSocials && /* @__PURE__ */ jsx("span", { className: "flex-1" }),
            themeToggle && /* @__PURE__ */ jsxs(Fragment, { children: [
              /* @__PURE__ */ jsx(Small, { className: "opacity-50", children: themeLabel }),
              /* @__PURE__ */ jsx(ThemeToggle, {})
            ] })
          ] })
        ] })
      }
    ) })
  ] });
}
function BrandCell({ brand, href, LinkComponent }) {
  return isExternal(href) ? /* @__PURE__ */ jsx(Cell, { href, ...EXTERNAL_REL, as: "a", children: brand }) : /* @__PURE__ */ jsx(Cell, { as: LinkComponent, href, children: brand });
}
function BrandLink({ brand, href, LinkComponent, onClick }) {
  if (isExternal(href)) {
    return /* @__PURE__ */ jsx("a", { href, onClick, ...EXTERNAL_REL, children: brand });
  }
  return createElement(
    LinkComponent,
    { href, onClick },
    brand
  );
}
function NavCell({ link, LinkComponent, scramble }) {
  const ref = useRef(null);
  const isExt = link.external ?? isExternal(link.href);
  const inner = /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsxs(Small, { children: [
      scramble ? /* @__PURE__ */ jsx(Scramble, { target: ref, children: link.label }) : link.label,
      /* @__PURE__ */ jsx(Blink, {})
    ] }),
    /* @__PURE__ */ jsx(HoverBg, {})
  ] });
  if (isExt) {
    return /* @__PURE__ */ jsx(
      Cell,
      {
        as: "a",
        className: "group relative cursor-pointer",
        href: link.href,
        onClick: link.onClick,
        ref,
        ...EXTERNAL_REL,
        children: inner
      }
    );
  }
  return /* @__PURE__ */ jsx(
    Cell,
    {
      as: LinkComponent,
      className: "group relative cursor-pointer",
      href: link.href,
      onClick: link.onClick,
      ref,
      children: inner
    }
  );
}
function MobileNavLink({
  link,
  LinkComponent,
  onNavigate,
  scramble
}) {
  const ref = useRef(null);
  const isExt = link.external ?? isExternal(link.href);
  const className = cn(
    "group relative flex cursor-pointer items-center border-b border-current/20 p-4"
  );
  const onClick = (e) => {
    link.onClick?.(e);
    onNavigate();
  };
  const children = /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsxs(Small, { children: [
      scramble ? /* @__PURE__ */ jsx(Scramble, { target: ref, children: link.label }) : link.label,
      /* @__PURE__ */ jsx(Blink, {})
    ] }),
    /* @__PURE__ */ jsx(HoverBg, {})
  ] });
  if (isExt) {
    return /* @__PURE__ */ jsx(
      "a",
      {
        className,
        href: link.href,
        onClick,
        ref,
        ...EXTERNAL_REL,
        children
      }
    );
  }
  return createElement(
    LinkComponent,
    { className, href: link.href, onClick, ref },
    children
  );
}
const EXTERNAL_REL = {
  rel: "noopener noreferrer",
  target: "_blank"
};
const isExternal = (href) => /^(https?:|mailto:|tel:)/i.test(href);
