// @nous-research/ui/dist/ui/components/tier-card.js
"use client";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { cn } from "../../utils/index.js";
import { ImageDistortion } from "./image-distortion.js";
import { Typography } from "./typography/index.js";
export function TierCard({
  badge,
  bullets,
  className,
  image,
  isCurrent = false,
  onSelect,
  overlay,
  price,
  selected = false,
  tint,
  tintStrength,
  title
}) {
  return /* @__PURE__ */ jsxs(
    "button",
    {
      className: cn(
        "group relative flex w-full cursor-pointer flex-col border border-current/20",
        "text-left transition-colors duration-300",
        selected && "border-midground/60",
        isCurrent && !selected && "border-midground/30",
        className
      ),
      onClick: onSelect,
      type: "button",
      children: [
        /* @__PURE__ */ jsx(
          "span",
          {
            "aria-hidden": true,
            className: cn(
              "arc-border transition-opacity duration-200",
              selected ? "opacity-100" : "opacity-0 group-hover:opacity-100"
            )
          }
        ),
        /* @__PURE__ */ jsxs(
          "div",
          {
            className: "relative aspect-[3/4] min-h-0 w-full flex-1 overflow-hidden",
            style: { backgroundColor: "var(--background)" },
            children: [
              /* @__PURE__ */ jsx(
                ImageDistortion,
                {
                  active: selected,
                  src: image,
                  tint,
                  tintStrength
                }
              ),
              overlay && /* @__PURE__ */ jsx(
                "div",
                {
                  className: "pointer-events-none absolute inset-0",
                  style: { backgroundColor: overlay, mixBlendMode: "color" }
                }
              ),
              /* @__PURE__ */ jsxs("div", { className: "pointer-events-none absolute inset-0 z-[1] flex flex-col justify-between p-3", children: [
                /* @__PURE__ */ jsxs("div", { className: "flex flex-col gap-0.5", children: [
                  /* @__PURE__ */ jsxs(
                    Typography,
                    {
                      variant: "sm",
                      className: cn(
                        "block drop-shadow-[0_1px_2px_rgba(0,0,0,0.5)] text-[1.2rem]",
                        "transition-colors",
                        selected && "text-midground"
                      ),
                      style: selected ? { mixBlendMode: "plus-lighter" } : void 0,
                      children: [
                        title,
                        badge && /* @__PURE__ */ jsx("span", { className: "ml-1 opacity-50", children: badge })
                      ]
                    }
                  ),
                  price.secondary ? /* @__PURE__ */ jsxs(Fragment, { children: [
                    /* @__PURE__ */ jsxs(
                      Typography,
                      {
                        className: "block text-md line-through opacity-50 drop-shadow-[0_1px_3px_rgba(0,0,0,0.6)]",
                        expanded: true,
                        style: { mixBlendMode: "plus-lighter" },
                        children: [
                          price.secondary,
                          price.secondarySuffix && /* @__PURE__ */ jsx("span", { className: "text-[1rem]", children: price.secondarySuffix })
                        ]
                      }
                    ),
                    /* @__PURE__ */ jsxs(
                      Typography,
                      {
                        className: "block text-xl font-bold drop-shadow-[0_1px_3px_rgba(0,0,0,0.6)]",
                        expanded: true,
                        style: { mixBlendMode: "plus-lighter" },
                        children: [
                          price.primary,
                          price.primarySuffix && /* @__PURE__ */ jsxs("span", { className: "text-[1rem] opacity-60", children: [
                            " ",
                            price.primarySuffix
                          ] })
                        ]
                      }
                    )
                  ] }) : /* @__PURE__ */ jsxs(
                    Typography,
                    {
                      className: "block text-xl font-bold drop-shadow-[0_1px_3px_rgba(0,0,0,0.6)]",
                      expanded: true,
                      style: { mixBlendMode: "plus-lighter" },
                      children: [
                        price.primary,
                        price.primarySuffix && /* @__PURE__ */ jsx("span", { className: "text-[1rem] opacity-60", children: price.primarySuffix })
                      ]
                    }
                  )
                ] }),
                bullets.length > 0 && /* @__PURE__ */ jsx("ul", { className: "flex flex-col gap-1", children: bullets.map((bullet, i) => /* @__PURE__ */ jsxs(
                  "li",
                  {
                    className: cn(
                      "font-courier text-display text-[1rem] leading-tight tracking-tight",
                      "drop-shadow-[0_1px_2px_rgba(0,0,0,0.5)]"
                    ),
                    children: [
                      "\xB7 ",
                      bullet
                    ]
                  },
                  typeof bullet === "string" ? bullet : i
                )) })
              ] })
            ]
          }
        )
      ]
    }
  );
}
