// @nous-research/ui/dist/ui/components/overlays/index.js
"use client";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { Glitch } from "./glitch.js";
import { Greys } from "./greys.js";
import { Lens } from "./lens-layers.js";
import { Noise } from "./noise.js";
import { Vignette } from "./vignette.js";
export { BLEND_MODES } from "./blend-modes.js";
export { Glitch } from "./glitch.js";
export { Greys } from "./greys.js";
export { Lens } from "./lens-layers.js";
export { Noise } from "./noise.js";
export { Vignette } from "./vignette.js";
export {
  $lightMode,
  applyLens,
  lens0,
  lens5i,
  LENS_0,
  LENS_5I,
  LENSES,
  toggleLens
} from "./lens";
const LAYER = "pointer-events-none fixed inset-0";
export function Overlays({ dark, initial }) {
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(Lens, { dark, initial }),
    /* @__PURE__ */ jsx(Noise, { className: LAYER, style: { zIndex: 101 } }),
    /* @__PURE__ */ jsx(Vignette, { className: LAYER, style: { zIndex: 99 } }),
    /* @__PURE__ */ jsx(Greys, { className: LAYER, style: { zIndex: 200 } }),
    /* @__PURE__ */ jsx(Glitch, { className: LAYER, style: { zIndex: 201 } })
  ] });
}
